✨ Initial implementation of Edge syncer

yana1205 opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary
This is the initial implementation of Edge syncer required for prototype of C2P on KCP.
 
## Related issue(s)

Fixes #

MikeSpreitzer commented:
BTW, #255 has merged now, this PR can be rebased to pick that up.

MikeSpreitzer commented on pkg/syncer/README.md:
````diff
@@ -0,0 +1,225 @@
+    $ kubectl kcp ws root:edge
+    Current workspace is "root:edge".
+    ```
+1. Create APIResourceSchema and APIExpose for PolicyReport CRD
````
s/APIExpose/APIExport/

yana1205 replied:
Thanks.
Fixed.

MikeSpreitzer commented on pkg/syncer/README.md:
````diff
@@ -0,0 +1,225 @@
+    ```
+1. Denature PolicyReport CRD in Kyverno Helm chart
+    ```
+    Name of CRD: policyreports.wgpolicyk8s.io => policyreports.wgpolicyk8s.io.denatured
+    Group in CRD: wgpolicyk8s.io.denatured => wgpolicyk8s.io
````
I do not understand exactly what is being said here.  Is this an English representation of some editing activity?

yana1205 replied:
My description was poor and wrong. I would express the following replacements in CustomResourceDefinition manifest for policyreport resource definition:
1. Replace `metadata.name: policyreports.wgpolicyk8s.io` with `metadata.name: policyreports.wgpolicyk8s.io.denatured`
2. Replace `spec.group: wgpolicyk8s.io` with `spec.group: wgpolicyk8s.io.denatured`

I will update the README so. Thanks for the review.

MikeSpreitzer commented on pkg/syncer/apis/edge/register.go:
```diff
@@ -0,0 +1,21 @@
+package edge
+
+const (
+	GroupName = "edge.kcp.io"
```
1. We keep a 1:1 correspondence between API group and source directory.
2. The API group source directories appear under https://github.com/kcp-dev/edge-mc/tree/main/pkg/apis .

yana1205 replied:
Thanks. I will move it to `/pkg/apis`. 

> for example, because you want to make a distinction between resources that are part of the interface of edge-mc vs. resources that are internal

Yes, this is also what I was wondering if syncer related APIs are part of edge-mc interface or not. EdgeSyncConfig can be referred by edge-mc controllers (PlacementTranslater?) if my understanding is correct. It makes sense for the API to be in the same api groups as edge-mc.

MikeSpreitzer commented on pkg/syncer/apis/edge/v1alpha1/edge-syncconfig.go:
```diff
@@ -0,0 +1,78 @@
+	Version string `json:"version,omitempty"`
+
+	// Name of down/up synced resource
+	Name string `json:"name,omitempty"`
```
When there is more than one object of a given kind in a given namespace, it would be more compact if a list of names could appear here.  If all objects of a given kind in a given namespace are desired, it would be more compact if `"*"` could appear here.

MikeSpreitzer commented on pkg/syncer/controller/controller.go:
```diff
@@ -0,0 +1,186 @@
+}
+
+// Start starts the controller workers.
+func (c *controller) Start(ctx context.Context, numThreads int) {
```
A better name for this method is "Run".  The word "start" does not imply waiting until completion, but "run" does.

MikeSpreitzer commented on pkg/syncer/controller/controller.go:
```diff
@@ -0,0 +1,186 @@
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
```
This also is a "run" rather than a "start".

yana1205 replied:
Thanks. Fixed.

MikeSpreitzer commented on pkg/syncer/controller/controller.go:
```diff
@@ -0,0 +1,186 @@
+	upSyncerError := c.upSyncer.ReInitializeClients(syncConfig.Spec.UpSyncedResources, syncConfig.Spec.Conversions)
+
+	if downsyncerError != nil && upSyncerError != nil {
+		return fmt.Errorf("failed to reinitialize downsyncer and upsyncer")
```
```suggestion
		return fmt.Errorf("failed to reinitialize downsyncer (%w) and upsyncer (%w)", downsyncerError, upSyncerError)
```

MikeSpreitzer commented on pkg/syncer/apis/edge/v1alpha1/edge-syncconfig.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Resource to be renatured
+type EdgeSynConversion struct {
```
Some sticklers may ask you to put the missing "c" in that name, but I will not.

clubanderson replied:
I would add a "c" here

MikeSpreitzer commented on pkg/syncer/controller/controller_test.go:
```diff
@@ -0,0 +1,122 @@
+			syncConfigInformerFactory.Start(ctx.Done())
+
+			err = nil
+			require.Eventually(t, func() bool {
+				syncConfig, _err := syncConfigInformer.Lister().Get(tc.syncConfig.Name)
+				if _err != nil {
+					if errors.IsNotFound(_err) {
+						return false
+					} else {
+						err = _err
+						return true
+					}
+				}
+				return syncConfig != nil
+			}, wait.ForeverTestTimeout, 1*time.Second)
```
Why not do the usual wait for `syncConfigInformer.HasSynced()` here?

MikeSpreitzer commented on pkg/syncer/controller/controller_test.go:
```diff
@@ -0,0 +1,122 @@
+			}, wait.ForeverTestTimeout, 1*time.Second)
+			assert.NoError(t, err)
+
+			controller.Start(ctx, 1)
```
Because this method waits until the context is closed, this test hangs at this point.

yana1205 replied:
Thanks. I was in a middle of writing the test. I have just fixed it and completed to write simple test cases.

MikeSpreitzer commented on pkg/syncer/controller/controller_test.go:
```diff
@@ -0,0 +1,122 @@
+			assert.NoError(t, err)
+
+			controller.Start(ctx, 1)
+			require.Eventually(t, func() bool {
```
If the previous statement did not hang, this one would.  It looks like the test is not fully written yet.  That's OK.

yana1205 replied:
Resolved in the above fix.

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+	"time"
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/syncer/apis/edge/v1alpha1"
+	kcpclientset "github.com/kcp-dev/edge-mc/pkg/syncer/client/clientset/versioned"
```
I think it would be clearer if the name of this import used "syncer" in place of "kcp".  Same for the informers import.

MikeSpreitzer commented on pkg/syncer/apis/edge/v1alpha1/edge-syncconfig.go:
```diff
@@ -0,0 +1,78 @@
+// +genclient:nonNamespaced
+// +kubebuilder:resource:scope=Cluster,shortName=esc
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type EdgeSyncConfig struct {
```
BTW, you do not need "Edge" in the name of this type, it is not going in the same package as a TMC or any other syncer config.

yana1205 replied:
Thanks. I will do it when I consolidate syncer apis into `/pkg/apis`.

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+			listOptions.FieldSelector = fields.OneTermEqualSelector("metadata.name", cfg.SyncTargetName).String()
+		},
+	))
+	syncConfigInformer := syncConfigInformerFactory.Edge().V1alpha1().EdgeSyncConfigs()
```
https://github.com/kcp-dev/edge-mc/pull/206#discussion_r1128492162

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+	))
+	syncConfigInformer := syncConfigInformerFactory.Edge().V1alpha1().EdgeSyncConfigs()
+
+	syncConfigInformer.Lister().List(labels.Everything()) // TODO: Remove (for now, need to invoke List at once)
```
Why?  What difference does this make?

yana1205 replied:
Without this line, nothing comes to event handlers even when I add/update/delete EdgeSyncConfig resource in k8s. I am still debugging what is the issue.

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+	syncConfigInformerFactory.WaitForCacheSync(ctx.Done())
+
+	upstreamConfig := rest.CopyConfig(cfg.UpstreamConfig)
+	rest.AddUserAgent(upstreamConfig, "edge-mc#syncer/"+kcpVersion)
```
This makes `upstreamConfig` the same as `bootstrapConfig`, right?  Why duplicate this code?

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+		return err
+	}
+	downstreamDiscoveryClient := discovery.NewDiscoveryClientForConfigOrDie(downstreamConfig)
+	downstreamClientFactory, err := syncers.NewClientFactory(logger, downstreamDyClient, downstreamDiscoveryClient)
```
"Dy" does not make a great contrast with "Discovery".  Replacing "Dy" with "Dyn" or "Dynamic" would be clearer.

MikeSpreitzer commented on pkg/syncer/syncer.go:
```diff
@@ -0,0 +1,153 @@
+	return nil
+}
+
+func startSync(ctx context.Context, logger klog.Logger, cfg *SyncerConfig, syncConfigLister edgev1alpha1listers.EdgeSyncConfigLister, upSyncer *syncers.UpSyncer, downSyncer *syncers.DownSyncer) {
```
This also is "run" rather than "start".

MikeSpreitzer commented on pkg/syncer/syncers/client.go:
```diff
@@ -0,0 +1,83 @@
+	var createdObj *unstructured.Unstructured
+	var err error
+	if c.IsNamespaced() {
+		createdObj, err = c.ResourceClient.Namespace(resource.Namespace).Create(context.Background(), unstObj, v1.CreateOptions{})
```
I recommend filling in the `FieldManager` value for the various meta/v1/XXXOptions

MikeSpreitzer commented on pkg/syncer/syncers/client.go:
```diff
@@ -0,0 +1,83 @@
+		updatedObj, err = c.ResourceClient.UpdateStatus(context.Background(), unstObj, v1.UpdateOptions{})
+	}
+	return updatedObj, err
+}
```
No Delete?

yana1205 replied:
Right. Deletion is not cared for the 1st implementation.

MikeSpreitzer commented on pkg/syncer/apis/edge/v1alpha1/edge-syncconfig.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Resource specifies down/up synced resource with exact GVK and name (and namespace if not cluster scoped resource)
+type EdgeSyncConfigResource struct {
```
FYI this is not a typical usage of the word "Resource" in the world of Kubernetes code.  While in the wide world of the web, "resource" is anything that has a URL or URI, in Kubernetes code the word "resource" usually has a more specific meaning: a kind of object.  That is, a collection of objects rather than a single object.  Of course, the code base is not entirely consistent in this regard.  The public-facing documentation is worse, it is definitely mixed-up in this regard.

I recommend using the term "object" in the code when an individual API object is meant, reserving "resource" for a kind of object.

Here is one place where the rubber meets the road.  As you know, "kind" and "resource" are two equivalent but distinct way of identifying a sort of object.  In the XXXResource struct at hand, one would expect that the Kind and Name fields hold the kind and resource, respectively.  For example, as in https://github.com/kubernetes/apimachinery/blob/v0.26.3/pkg/apis/meta/v1/types.go#L1081 .  But that is not what is going on here.

MikeSpreitzer commented on pkg/syncer/apis/edge/v1alpha1/edge-syncconfig.go:
```diff
@@ -0,0 +1,78 @@
+// Resource specifies down/up synced resource with exact GVK and name (and namespace if not cluster scoped resource)
+type EdgeSyncConfigResource struct {
+	// Kind of down/up synced resource
+	Kind string `json:"kind,omitempty"`
```
The implementation of the syncer would be simpler if this field held the resource rather than the kind.  For example, "deployments" rather than "Deployment".  That would remove the need to go through "rest mapping" to get from kind to resource.

yana1205 replied:
Yes, I agree with. Actually, I had to write the code to go through rest mapping. Though, as a user who creates EdgeSyncConfig, I am wondering if "resource" is always represented as lower case of "kind" with "s" appended (; `strings.ToLower(kind) + "s"`). Actually, I can find "kind" in k8s yaml manifest but not find "resource". Maybe it's responsible of a user who creates EdgeSyncConfig with correct "resource" by somehow finding it from somewhere.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func (ds *DownSyncer) SyncOne(resource edgev1alpha1.EdgeSyncConfigResource, conversions []edgev1alpha1.EdgeSynConversion) error {
+	ds.logger.V(3).Info(fmt.Sprintf("downsync '%s'", resourceToString(resource)))
```
If you use `%q` instead of `%s` then you do not have to supply the quotes in the source and you also get the usual escaping for special characters.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			ds.logger.V(3).Info(fmt.Sprintf("  not found '%s' in upstream", resourceToString(resourceForUp)))
+			ds.logger.V(3).Info(fmt.Sprintf("  skip downsync '%s'", resourceToString(resourceForUp)))
```
Should there be a downstream delete here?

yana1205 replied:
Yes, I will create a gitissue for deletion support.

yana1205 replied:
I will create a gitissue for deletion support.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// create
+			ds.logger.V(3).Info(fmt.Sprintf("  create '%s' in upstream since it's not found", resourceToString(resourceForDown)))
```
Isn't the creation being done in the downstream?

yana1205 replied:
Yes, correct. Will fix the comment.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+	} else {
+		if downstreamResource != nil {
+			// update
+			ds.logger.V(3).Info(fmt.Sprintf("  update '%s' in upstream since it's found", resourceToString(resourceForDown)))
```
The update is in the downstream, right?

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+			upstreamResource.SetUID("")
+			upstreamResource.SetManagedFields(nil)
+			applyConversion(upstreamResource, resourceForDown)
+			if _, err := downstreamClient.Update(resourceForDown, upstreamResource); err != nil {
```
You could save significant runtime costs if this were conditional on whether there is actually a difference to be made here.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+			applyConversion(upstreamResource, resourceForDown)
+			if _, err := downstreamClient.Create(resourceForDown, upstreamResource); err != nil {
+				ds.logger.Error(err, fmt.Sprintf("failed to create resource to downstream '%s'", resourceToString(resourceForDown)))
+				return err
```
It would be great to return this information upstream.  In my thinking about summarization, I am hoping to be able to summarize it.  You can see my first draft proposal for summarization in https://docs.google.com/document/d/1yY-XgOtCnOPKpVdxdPVhH5ixRN6IWQQe0s3lip1ID0Y (view it as a member of kcp-dev), and I am in the midst of creating a second draft to propose to edge-mc.  Do not worry about it in this PR.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+			ds.logger.V(3).Info(fmt.Sprintf("  skip status upsync '%s'", resourceToString(resourceForDown)))
+			return nil
+		} else {
+			ds.logger.Error(err, fmt.Sprintf("failed to get resource from upstream '%s'", resourceToString(resourceForDown)))
```
you mean from downstream here, right?

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+	status, found, err := unstructured.NestedMap(downstreamResource.Object, "status")
+	if err != nil {
+		ds.logger.Error(err, fmt.Sprintf("failed to extract status from downstream object '%s'", resourceToString(resourceForDown)))
+		return err
```
Do you want a retry in this case?

yana1205 replied:
No... Maybe it would be better to have a flag in member variable of a resource client (/pkg/syncer/syncers/client.go) to indicate whether a resource has valid "status" field or not.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+		return err
+	}
+	upstreamResource.Object["status"] = status
+	updatedResource := unstructured.Unstructured{
```
Why is this copy being made?

yana1205 replied:
Maybe this has been left unintentionally. I confirmed that the status is properly upsynced without this copy. Thanks. Will be fixed in this PR.

MikeSpreitzer commented on pkg/syncer/syncers/downsyncer.go:
```diff
@@ -0,0 +1,181 @@
+	updatedResource := unstructured.Unstructured{
+		Object: upstreamResource.Object,
+	}
+	applyConversion(upstreamResource, resourceForUp)
```
Did you mean to apply conversion to `updatedResource` ?

MikeSpreitzer commented on pkg/syncer/syncers/upsyncer.go:
```diff
@@ -0,0 +1,128 @@
+		return err
+	}
+	resourceForDown := convertToUpstream(resource, conversions)
+	us.logger.V(3).Info(fmt.Sprintf("  get '%s' from downstream", resourceToString(resourceForDown)))
```
Structured logging?  If not, `%q`?

MikeSpreitzer commented on pkg/syncer/syncers/upsyncer.go:
```diff
@@ -0,0 +1,128 @@
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			us.logger.V(3).Info(fmt.Sprintf("  not found '%s' in downstream", resourceToString(resourceForDown)))
+			us.logger.V(3).Info(fmt.Sprintf("  skip upsync '%s'", resourceToString(resourceForDown)))
```
Delete?

yana1205 replied:
Yes, should be. Will be covered in https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1146017961

MikeSpreitzer commented on pkg/syncer/syncers/upsyncer.go:
```diff
@@ -0,0 +1,128 @@
+		us.logger.Error(err, fmt.Sprintf("failed to get client '%s'", resourceToString(resource)))
+		return err
+	}
+	resourceForDown := convertToUpstream(resource, conversions)
```
This is actually for up not down, right?

yana1205 commented on cmd/syncer/helm/converter.go:
```diff
@@ -0,0 +1,192 @@
+
+	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/syncer/apis/edge/v1alpha1"
+)
+
```
No, there isn't. I will fill the command description and update README in this PR. This command is not related with Syncer but just an utility to generate EdgeSyncConfig manifest from yaml manifests returned by Helm (`helm template kyverno --dry-run`). I would like to leave this command.

MikeSpreitzer replied:
FYI, two smaller PRs are easier to review than one bigger PR.

yana1205 replied:
Yes, I totally agree with that. I will do that from the next.

MikeSpreitzer replied:
Today I see no reference to this command.

```console
(venv) mspreitz@mjs12 kubestellar % find * \( -path docs/venv -prune \) -or \( -name "*.go" -exec grep helm \{\} \; -print -exec echo \; \)
	fs := pflag.NewFlagSet("helmconverter", pflag.ExitOnError)
	fs.StringVar(&pathToHelmTemplate, "path-to-helm-template", "", "path to Helm tamplate file")
cmd/syncer/helm/converter.go

(venv) mspreitz@mjs12 kubestellar % 
```

It is a problem to build a variant executable named `helm` and put it in the KubeStellar `bin` directory because that will shadow the regular `helm` for users of KubeStellar.

MikeSpreitzer replied:
Today I can not find any documentation of the use of this command.

MikeSpreitzer commented:
BTW, I agree with the approach taken here for how to denature: change the API group from `some.original.thing` to `some.original.thing.denatured`.

MikeSpreitzer commented:
There is a bit of lint to pick here before this can merge.

yana1205 commented:
I fixed the codes following linter message and now PR checks passed. 

> BTW, https://github.com/kcp-dev/edge-mc/pull/255 has merged now, this PR can be rebased to pick that up.

I force-pushed with the PR branch rebased from the top of main.

I added thumsup mark in the comment that I fixed in this PR. For things not included in this PR, I summarized them as followup items.
- Move up syncer apis to /pkg/apis
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145692371
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145826529
- Support a list of names in EdgeSyncConfig
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145592961
- Bug? Need to identify why informer doesn't inform anything without invoking List or Get once
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145840080
- Support propagation of deletion
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145921774
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1146017961
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1146058283
- Use Resource rather than Kind to make logic simpler
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145644407
- When syncer failed to create/update resources in downstream, send it back to upstream 
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145656244
- In order to reduce computation cost, the information if a Resource type doesn not have status field or not can be cached.
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1146035011
- Reduce computational cost by introducing object comparison between up/down before doing update
  - https://github.com/kcp-dev/edge-mc/pull/259#discussion_r1145652241

MikeSpreitzer commented on cmd/syncer/helm/converter.go:
```diff
@@ -55,7 +56,7 @@ func main() {
 	_ = ctx
 	fs.VisitAll(func(flg *pflag.Flag) {
-		logger.V(1).Info("Command line flag", flg.Name, flg.Value)
+		logger.V(1).Info(fmt.Sprintf("Command line flag %s %s", flg.Name, flg.Value))
```
Sadly, the linter is overly picky about this.  You can tell because the linter output in this case explicitly tells you that there is a list of accepted complaints that you can update; see https://github.com/kcp-dev/edge-mc/blob/4219d05658f25bc36f0d01d4b57f8266bd566f2c/hack/verify-contextual-logging.sh#L63.  See the exceptions already listed in https://github.com/kcp-dev/edge-mc/blob/4219d05658f25bc36f0d01d4b57f8266bd566f2c/hack/logcheck.out .
